edm4hep.yaml

```
@@ -780,6 +780,36 @@ datatypes:
     OneToOneRelations:
       - edm4hep::Track track           // the corresponding track
 
+
+  #===== Generator related data =====
+
+  #---------- GeneratorParameters
+  edm4hep::GeneratorParameters:
+    Description: "Generator parameters"
+    Author: "EDM4hep authors"
+    Members:
+      - double event_scale              // event scale
+      - double alphaQED                 // alpha_QED
+      - double alphaQCD                 // alpha_QCD
+      - int signal_process_id           // id of signal process
+      - double sqrt_s  [GeV]            // sqrt(s)
+    VectorMembers:
+      - double crossSections            // list of cross sections in pb
+      - double crossSectionErrors       // list of cross section errors
+
+  #---------- GeneratorPdfInfo
+  edm4hep::GeneratorPdfInfo:
+    Description: "Generator pdf information"
+    Author: "EDM4hep authors"
+    Members:
+      - std::array<int, 2> parton_id      // Parton PDG id
+      - std::array<int, 2> id             // LHAPDF PDF id (see https://lhapdf.hepforge.org/pdfsets.html)
+      - std::array<double, 2> x           // Parton momentum fraction
+      - std::array<double, 2> xf          // PDF value
+      - double scale [GeV]                // Factorisation scale
+    OneToManyRelations:
+      - edm4hep::MCParticle signal_vertex // pointing into MC Generators collection
```

```suggestion
      - edm4hep::MCParticle signal_vertex // pointing into MCParticle collection
```
? 
Or what is a MC Generators collection?

And what does this mean anyway? Because it being a relation already means that it points to a MCParticle collection. So what does `signal_vertex` actually mean?

---

Hi @andresailer,
since HepMC separates particles and vertices, they can identify the signal vertex by a a number. The use is to separate the hard scatter vertex (signal_vertex) to differentiate from overlaid pileup events, so even after merging one has a chance to separate signal from background/. 
So I guess it is used as a list of all particles belonging to this particular vertex?

Wondering (I had overlooked that it is part of pdfInfo): shouldn't this list/onetomany be part of the GeneratorParameters (rather than PDFINFO)?

Cheers,
Dirk

---

So this would be the `List of MCParticles that were generated by this Generator`?

---

sorry, I had not checked the HepMC definition, so "belonging" is very vague:
from the HepMC documentation: "V vertex OID status (comma-separated list of incoming OIDs)."
so in that spirit I would write it as the list of incoming particles of this particular vertex.

---

@hegner We definitely need a better docstring for this line

@dirkzerwas We don't have vertices in the HepMC3 sense. So can you give an example of what signal_vertex is for some example event?

---

@andresailer - yes. Indeed. We'll wait for @dirkzerwas to have a good example so that we write the proper thing

---

@andresailer 
here is an example from SHERPA. Process ZH@250GeV (I removed some stuff
HepMC::Version 3.02.07
HepMC::Asciiv3-START_EVENT_LISTING
T SHERPA-MC\|2.2.15\|Cho Oyu
E 0 7 16
U GEV MM
W 2.3990527749323822814809e-01 2.9672764031814940783455e-04 1.2613634180205666179830e+00 1.0000000000000000000000e+00 0.0000000000000000000000e+00
A 0 GenCrossSection 2.39905277e-01 2.39905277e-01 -1 -1
A 0 GenPdfInfo 11 -11 9.99988759e-01 9.22321388e-01 2.40092563e+02 1.13670760e+04 6.07635210e-01 0 0
A 0 alphaQCD 0.103428260737051
A 0 alphaQED 0.0077638411389137
A 0 event_scale 57644.4387826361
A 0 mpi -1
A 0 signal_process_vertex -5
...
P 1 0 11 0.000000000000e+00 0.000000000000e+00 1.250000000000e+02 1.250000000000e+02 0.000000000000e+00 4
P 2 1 11 0.000000000000e+00 0.000000000000e+00 1.250000000000e+02 1.250000000000e+02 0.000000000000e+00 11
P 3 0 -11 0.000000000000e+00 0.000000000000e+00 -1.250000000000e+02 1.250000000000e+02 0.000000000000e+00 4
P 4 3 -11 0.000000000000e+00 0.000000000000e+00 -1.250000000000e+02 1.250000000000e+02 0.000000000000e+00 11
P 5 4 -11 -0.000000000000e+00 -0.000000000000e+00 -1.152901735522e+02 1.152901735533e+02 5.110039371923e-04 11
P 6 4 22 0.000000000000e+00 0.000000000000e+00 -9.709826447783e+00 9.709826446650e+00 -1.482971707319e-04 1
P 7 2 11 -0.000000000000e+00 -0.000000000000e+00 1.249985948618e+02 1.249985948628e+02 5.110003775412e-04 11
P 8 2 22 0.000000000000e+00 0.000000000000e+00 1.405138223774e-03 1.405137179276e-03 -1.713279468285e-06 1
V -6 4 [5,7]
P 9 -6 11 0.000000000000e+00 0.000000000000e+00 1.249985948629e+02 1.249985948629e+02 1.907348632812e-06 3
P 10 -6 -11 0.000000000000e+00 0.000000000000e+00 -1.152901735533e+02 1.152901735533e+02 -2.336015455993e-06 3
V -5 1 [9,10]
...
so signal_vertex, as I read it, the important Vertex is "-5" V -5 refers to P 9 and 10 as incoming particles (electron, positron, but after ISR, so it is not the beam particles).
So if we want to keep the capability of moving seamless between HepMC and EDM4HEP, I would write the P9 and P10 particles into the OneToManyRelation.

Come to think of it: if we calculate the invariant mass of P9+P10 we get the stored event_scale. So I think it's a consistent definition (implemented by Sherpa at least).

---

sounds good to me!